src/peer.rs

```diff
@@ -197,17 +197,20 @@ pub(crate) fn new_peer<T: HandshakeHandler>(
                         return;
                     }
                     println!("Received data from peer: {:?}", data.len());
-                    let Ok(handler_id) = data[0..8].try_into() else {
-                        println!("Peer err received message couldn't determine handler id");
-                        continue;
-                    };
-                    let handler_id = u64::from_be_bytes(handler_id);
-                    let data = data[8..].to_vec();
-                    message_handlers
-                        .get_handler(handler_id)
-                        .unwrap()
-                        .send_message((peer_id.clone(), data))
-                        .unwrap();
+                    if let Err(err) = message_handler.deserialize_and_handle(&data, &peer_id) {
```

Comment:
So here, with the other comment about associated types, we would first deserialize (and check for errors), and then handle the message (and check for errors)

Reply:
We didn't go for associated types but splitted the functions